src/Dfc.CourseDirectory.WebV2/Features/Venues/AddVenue/PostcodeSearch.cs

```diff
@@ -0,0 +1,194 @@
+        IRequestHandler<SearchCommand, OneOf<ModelWithErrors<SearchCommand>, SearchViewModel>>,
+        IRequestHandler<SelectCommand, OneOf<ModelWithErrors<SearchViewModel>, Success>>
+    {
+        private readonly IAddressSearchService _addressSearchService;
+        private readonly ISearchClient<Onspd> _onspdSearchClient;
+        private readonly JourneyInstance<AddVenueJourneyModel> _journeyInstance;
+
+        public Handler(
+            IAddressSearchService addressSearchService,
+            ISearchClient<Onspd> onspdSearchClient,
+            JourneyInstanceProvider journeyInstanceProvider)
+        {
+            _addressSearchService = addressSearchService;
+            _onspdSearchClient = onspdSearchClient;
+            _journeyInstance = journeyInstanceProvider.GetInstance<AddVenueJourneyModel>();
+        }
+
+        public async Task<OneOf<ModelWithErrors<SearchCommand>, SearchViewModel>> Handle(
+            SearchCommand request,
+            CancellationToken cancellationToken)
+        {
+            ThrowIfFlowStateNotValid();
+
+            var validator = new SearchCommandValidator();
+            var validationResult = validator.Validate(request);
+
+            if (!validationResult.IsValid)
+            {
+                return new ModelWithErrors<SearchCommand>(request, validationResult);
+            }
+
+            // We need the postcode to exist in ONSPD so we can resolve a lat/lng.
+            // Check it's present before we do an address search so we can output an error early.
+
+            var onspdSearchResult = await _onspdSearchClient.Search(new OnspdSearchQuery() { Postcode = request.Postcode });
+
+            if (onspdSearchResult.Items.Count == 0)
+            {
+                return CreateInvalidPostcodeResponse();
+            }
+
+            var vm = await CreateViewModel(request.Postcode);
+
+            if (vm.Results.Count == 0)
+            {
+                return CreateInvalidPostcodeResponse();
+            }
+
+            return vm;
+
+            ModelWithErrors<SearchCommand> CreateInvalidPostcodeResponse() =>
+                new ModelWithErrors<SearchCommand>(
+                    request,
+                    new ValidationResult(new[]
+                    {
+                        new ValidationFailure(nameof(SearchCommand.Postcode), "Enter a real postcode")
+                    }));
+        }
+
+        public async Task<OneOf<ModelWithErrors<SearchViewModel>, Success>> Handle(
+            SelectCommand request,
+            CancellationToken cancellationToken)
+        {
+            ThrowIfFlowStateNotValid();
+
+            var validator = new SelectCommandValidator();
+            var validationResult = validator.Validate(request);
+
+            if (!validationResult.IsValid)
+            {
+                // No address specified. Re-run the search with a prompt to select an address.
+                var vm = await CreateViewModel(request.Postcode);
```

Comment:
Potential bug - why would you want to re-run the search and recover in this circumstance? Especially without having conducted the onspd postcode verification?

Reply:
Amended with latest commit.